--- linux-3.10.0-0.rc0.git15.1.fc20.x86_64/drivers/mmc/host/sdhci-tegra.c.orig	2013-05-05 16:02:40.473147107 +0100
+++ linux-3.10.0-0.rc0.git15.1.fc20.x86_64/drivers/mmc/host/sdhci-tegra.c	2013-05-05 16:03:08.153399199 +0100
@@ -214,7 +214,7 @@
 #endif
 	{}
 };
-MODULE_DEVICE_TABLE(of, sdhci_dt_ids);
+MODULE_DEVICE_TABLE(of, sdhci_tegra_dt_match);
 
 static void sdhci_tegra_parse_dt(struct device *dev,
 					struct sdhci_tegra *tegra_host)
