diff -urNp linux-3.9.4-300.fc19.armv7hl_orig/drivers/video/omap2/dss/core.c linux-3.9.4-300.fc19.armv7hl/drivers/video/omap2/dss/core.c
--- linux-3.9.4-300.fc19.armv7hl_orig/drivers/video/omap2/dss/core.c	2013-04-28 20:36:01.000000000 -0400
+++ linux-3.9.4-300.fc19.armv7hl/drivers/video/omap2/dss/core.c	2013-05-31 12:24:07.711334359 -0400
@@ -596,6 +596,9 @@ static int __init omap_dss_init(void)
 {
 	int r;
 
+	/* hack to load panel-tfp410 driver */
+	request_module("panel-tfp410");
+
 	r = omap_dss_bus_register();
 	if (r)
 		return r;
Binary files linux-3.9.4-300.fc19.armv7hl_orig/drivers/video/omap2/dss/.Makefile.swp and linux-3.9.4-300.fc19.armv7hl/drivers/video/omap2/dss/.Makefile.swp differ
