diff --git a/sound/soc/samsung/dma.c b/sound/soc/samsung/dma.c
index 21b7926..19e6662 100644
--- a/sound/soc/samsung/dma.c
+++ b/sound/soc/samsung/dma.c
@@ -76,7 +76,7 @@ static void dma_enqueue(struct snd_pcm_substream *substream)
 
 	pr_debug("Entered %s\n", __func__);
 
-	limit = (prtd->dma_end - prtd->dma_start) / prtd->dma_period;
+	limit = (u32)(prtd->dma_end - prtd->dma_start) / prtd->dma_period;
 
 	pr_debug("%s: loaded %d, limit %d\n",
 				__func__, prtd->dma_loaded, limit);
