--- a/kernel/sched/bfs.c	2014-04-08 16:11:44.996610427 +0400
+++ b/kernel/sched/bfs.c	2014-04-08 16:11:51.897722801 +0400
@@ -243,11 +243,6 @@
 #define this_rq()		(&__get_cpu_var(runqueues))
 #define task_rq(p)		cpu_rq(task_cpu(p))
 #define cpu_curr(cpu)		(cpu_rq(cpu)->curr)
-/*
- * sched_domains_mutex serialises calls to init_sched_domains,
- * detach_destroy_domains and partition_sched_domains.
- */
-static DEFINE_MUTEX(sched_domains_mutex);
 
 /*
  * By default the system creates a single root-domain with all cpus as
--- a/kernel/sched/bfs_sched.h	2014-04-08 16:14:27.697259879 +0400
+++ b/kernel/sched/bfs_sched.h	2014-04-08 16:14:38.570436950 +0400
@@ -99,6 +99,12 @@
 	return rq->clock_task;
 }
 
+/*
+ * sched_domains_mutex serialises calls to init_sched_domains,
+ * detach_destroy_domains and partition_sched_domains.
+ */
+static DEFINE_MUTEX(sched_domains_mutex);
+
 #define rcu_dereference_check_sched_domain(p) \
 	rcu_dereference_check((p), \
 			      lockdep_is_held(&sched_domains_mutex))
