diff -up compat-wireless-2011-12-18/config.mk.orig compat-wireless-2011-12-18/config.mk
--- compat-wireless-2011-12-18/config.mk.orig	2011-12-18 16:10:34.000000000 -0500
+++ compat-wireless-2011-12-18/config.mk	2011-12-19 15:54:24.962156028 -0500
@@ -14,6 +14,9 @@ else
 include $(KLIB_BUILD)/.config
 endif
 
+# Enable SSB module (needs to be visible in compat_autoconf.h)
+CONFIG_SSB=m
+
 ifneq ($(wildcard $(KLIB_BUILD)/Makefile),)
 
 COMPAT_LATEST_VERSION = 3
diff -up compat-wireless-2011-12-18/drivers/misc/eeprom/Makefile.orig compat-wireless-2011-12-18/drivers/misc/eeprom/Makefile
--- compat-wireless-2011-12-18/drivers/misc/eeprom/Makefile.orig	2011-12-18 16:10:36.000000000 -0500
+++ compat-wireless-2011-12-18/drivers/misc/eeprom/Makefile	2011-12-19 15:54:24.963156015 -0500
@@ -1 +1 @@
-obj-$(CONFIG_EEPROM_93CX6)	+= eeprom_93cx6.o
+#obj-$(CONFIG_EEPROM_93CX6)	+= eeprom_93cx6.o
diff -up compat-wireless-2011-12-18/drivers/net/ethernet/atheros/Makefile.orig compat-wireless-2011-12-18/drivers/net/ethernet/atheros/Makefile
--- compat-wireless-2011-12-18/drivers/net/ethernet/atheros/Makefile.orig	2011-12-18 16:10:35.000000000 -0500
+++ compat-wireless-2011-12-18/drivers/net/ethernet/atheros/Makefile	2011-12-19 15:54:24.963156015 -0500
@@ -2,7 +2,7 @@
 # Makefile for the Atheros network device drivers.
 #
 
-obj-$(CONFIG_ATL1) += atlx/
-obj-$(CONFIG_ATL2) += atlx/
-obj-$(CONFIG_ATL1E) += atl1e/
-obj-$(CONFIG_ATL1C) += atl1c/
+#obj-$(CONFIG_ATL1) += atlx/
+#obj-$(CONFIG_ATL2) += atlx/
+#obj-$(CONFIG_ATL1E) += atl1e/
+#obj-$(CONFIG_ATL1C) += atl1c/
diff -up compat-wireless-2011-12-18/drivers/net/usb/Makefile.orig compat-wireless-2011-12-18/drivers/net/usb/Makefile
--- compat-wireless-2011-12-18/drivers/net/usb/Makefile.orig	2011-12-18 16:10:36.000000000 -0500
+++ compat-wireless-2011-12-18/drivers/net/usb/Makefile	2011-12-19 15:54:24.964156002 -0500
@@ -2,7 +2,7 @@
 # Makefile for USB Network drivers
 #
 
-obj-$(CONFIG_USB_NET_COMPAT_CDCETHER)	+= cdc_ether.o
-obj-$(CONFIG_USB_NET_COMPAT_RNDIS_HOST)	+= rndis_host.o
-obj-$(CONFIG_USB_COMPAT_USBNET)	+= usbnet.o
+#obj-$(CONFIG_USB_NET_COMPAT_CDCETHER)	+= cdc_ether.o
+#obj-$(CONFIG_USB_NET_COMPAT_RNDIS_HOST)	+= rndis_host.o
+#obj-$(CONFIG_USB_COMPAT_USBNET)	+= usbnet.o
 
diff -up compat-wireless-2011-12-18/Makefile.orig compat-wireless-2011-12-18/Makefile
--- compat-wireless-2011-12-18/Makefile.orig	2011-12-18 16:10:34.000000000 -0500
+++ compat-wireless-2011-12-18/Makefile	2011-12-19 15:54:24.965155989 -0500
@@ -41,8 +41,8 @@ endif
 
 endif
 
-obj-$(CONFIG_COMPAT_BLUETOOTH) += net/bluetooth/
-obj-$(CONFIG_COMPAT_BLUETOOTH_MODULES) += drivers/bluetooth/
+#obj-$(CONFIG_COMPAT_BLUETOOTH) += net/bluetooth/
+#obj-$(CONFIG_COMPAT_BLUETOOTH_MODULES) += drivers/bluetooth/
 
 else
 
