From: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>
Date: Fri, 5 Jun 2015 14:37:34 +0200
Subject: [PATCH] kdbus: drop useless goto

Signed-off-by: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>
Reviewed-by: David Herrmann <dh.herrmann@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 ipc/kdbus/names.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/ipc/kdbus/names.c b/ipc/kdbus/names.c
index 5f5d84ea0e8e..d77ee08afeda 100644
--- a/ipc/kdbus/names.c
+++ b/ipc/kdbus/names.c
@@ -514,8 +514,6 @@ int kdbus_cmd_name_acquire(struct kdbus_conn *conn, void __user *argp)
 
 	ret = kdbus_name_acquire(conn->ep->bus->name_registry, conn, item_name,
 				 cmd->flags, &cmd->return_flags);
-	if (ret < 0)
-		goto exit_dec;
 
 exit_dec:
 	atomic_dec(&conn->name_count);
