--- a/kernel/sched/bfs.c	2015-03-19 21:11:38.189898144 +0300
+++ b/kernel/sched/bfs.c	2015-03-19 21:13:17.325550342 +0300
@@ -235,12 +235,6 @@
 DEFINE_PER_CPU_SHARED_ALIGNED(struct rq, runqueues);
 #ifdef CONFIG_SMP
 /*
- * sched_domains_mutex serialises calls to init_sched_domains,
- * detach_destroy_domains and partition_sched_domains.
- */
-static DEFINE_MUTEX(sched_domains_mutex);
-
-/*
  * By default the system creates a single root-domain with all cpus as
  * members (mimicking the global state we have today).
  */
--- a/kernel/sched/bfs_sched.h	2015-03-19 21:11:36.429868811 +0300
+++ b/kernel/sched/bfs_sched.h	2015-03-19 21:11:38.191898177 +0300
@@ -123,6 +123,12 @@
 };
 #endif
 
+/*
+ * sched_domains_mutex serialises calls to init_sched_domains,
+ * detach_destroy_domains and partition_sched_domains.
+ */
+static DEFINE_MUTEX(sched_domains_mutex);
+
 #define rcu_dereference_check_sched_domain(p) \
 	rcu_dereference_check((p), \
 			      lockdep_is_held(&sched_domains_mutex))
